File: matlab/lib/+Ice/RequestFailedException.m
classdef RequestFailedException < Ice.DispatchException
    %REQUESTFAILEDEXCEPTION The base class for the 3 NotExist exceptions.
    %
    %   RequestFailedException Properties:
    %     id - The identity of the Ice Object to which the request was sent.
    %     facet - The facet to which the request was sent.
    %     operation - The operation name of the request.

    % Copyright (c) ZeroC, Inc.

    properties
        %ID The identity of the Ice Object to which the request was sent.
        %   Ice.Identity scalar
        id (1, :) Ice.Identity = Ice.Identity.empty

        %FACET The facet to which the request was sent.
        %   character vector
        facet (1, :) char = ''

        %OPERATION The operation name of the request.
        %   character vector
        operation (1, :) char = ''
    end
    methods (Hidden)
        function obj = RequestFailedException(replyStatus, id, facet, operation, errID, what)
            if nargin == 0
                superArgs = {};
            else
                assert(nargin == 6, 'Invalid number of arguments');
                superArgs = {replyStatus, errID, what};
            end
            obj@Ice.DispatchException(superArgs{:});
            if nargin > 0
                obj.id = id;
                obj.facet = facet;
                obj.operation = operation;
            end
        end
    end
end

File: java/src/com.zeroc.ice/src/main/java/com/zeroc/Ice/RequestFailedException.java
// Copyright (c) ZeroC, Inc.

package com.zeroc.Ice;

/**
 * This exception is raised if a request failed. This exception, and all exceptions derived from
 * {@link DispatchException}, are transmitted by the Ice protocol.
 */
public abstract class RequestFailedException extends DispatchException {
    @Override
    public String ice_id() {
        return "::Ice::RequestFailedException";
    }

    /** The identity of the Ice Object to which the request was sent. */
    public final Identity id;

    /** The facet to which the request was sent. */
    public final String facet;

    /** The operation name of the request. */
    public final String operation;

    // Logically protected too.
    RequestFailedException(ReplyStatus replyStatus) {
        super(replyStatus.value());
        this.id = new Identity();
        this.facet = "";
        this.operation = "";
    }

    // Logically protected too.
    RequestFailedException(ReplyStatus replyStatus, Identity id, String facet, String operation) {
        super(replyStatus.value(), createMessage(replyStatus, id, facet, operation));
        this.id = id;
        this.facet = facet;
        this.operation = operation;
    }

    private static String createMessage(
            ReplyStatus replyStatus, Identity id, String facet, String operation) {
        return "Dispatch failed with "
            + replyStatus
            + " { id = '"
            + Util.identityToString(id)
            + "', facet = '"
            + facet
            + "', operation = '"
            + operation
            + "' }";
    }

    private static final long serialVersionUID = 4181164754424262091L;
}
